Use non-deprecated method of re-processing MatrixRTCSession keys

hughns opened the pull request:
As per https://github.com/matrix-org/matrix-js-sdk/pull/4423 the key can be undefined and so we should handle this rather than waiting for SubtleCrypto.importKey() to fail.

hughns committed:
Handle case of encryption key for an index to be undefined

As per https://github.com/matrix-org/matrix-js-sdk/pull/4423 the key can be undefined and so we should handle this rather than waiting for SubtleCrypto.importKey() to fail.

AndrewFerr commented on src/e2ee/matrixKeyProvider.ts:
```diff
@@ -41,7 +41,9 @@ export class MatrixKeyProvider extends BaseKeyProvider {
       encryptionKeys,
     ] of this.rtcSession.getEncryptionKeys()) {
       for (const [index, encryptionKey] of encryptionKeys.entries()) {
```
As it is currently written (and is sufficiently captured in type hints), the array returned by `getEncryptionKeys` is a [sparse array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Indexed_collections#sparse_arrays), which contains empty slots that are skipped by array iteration methods.

Thus, instead of using a `for...of` loop and having to check for `undefined`, this could simply use `forEach`:

```suggestion
      encryptionKeys.forEach((encryptionKey, index) => {
```

Note that this will work only if the array returned by `getEncryptionKeys` remains typed as `Array<Uint8Array>` instead of `Array<Uint8Array | undefined>`.

hughns replied:
Well spotted. 👍 However, this is now redundant given the refactor from https://github.com/matrix-org/matrix-js-sdk/pull/4427

hughns committed:
Add some test coverage whilst we are here

hughns committed:
Merge branch 'livekit' into hughns/undefined-encryption-key

hughns committed:
Use release version of matrix-js-sdk

Diff is https://github.com/matrix-org/matrix-js-sdk/compare/baa6d135065637c9769c61325c69709d3618f5f1...v34.7.0

hughns committed:
Use RTCSession. reemitEncryptionKeys()

hughns committed:
Add some test coverage whilst we are here

hughns committed:
Lint

toger5 commented on src/e2ee/matrixKeyProvider.test.ts:
```diff
@@ -0,0 +1,72 @@
+    expect(keyProvider).toBeTruthy();
+  });
+
+  test("listens for key requests and emits existing keys", () => {
+    const keyProvider = new MatrixKeyProvider();
+
+    const session = mockRTCSession();
+
+    keyProvider.setRTCSession(session);
+
+    expect(session.on).toHaveBeenCalledWith(
+      MatrixRTCSessionEvent.EncryptionKeyChanged,
+      expect.any(Function),
+    );
+    expect(session.off).not.toHaveBeenCalled();
+  });
```
I suppose we test that updating session (join/leave/receive new keys) also trigger a session `EncryptionKeyChanged` emit somewhere else?

hughns replied:
Yes, these would be at https://github.com/matrix-org/matrix-js-sdk/blob/develop/spec/unit/matrixrtc/MatrixRTCSession.spec.ts

toger5 reviewed (approved):
Very nice this has gotten simpler.

Added a comment since I am not too sure about our E2EE test coverage.
